pallets/payment/src/lib.rs

```diff
@@ -422,95 +445,58 @@ pub mod pallet {
+		/// For cancelling a payment, recipient_share = 0
+		/// For releasing a payment, recipient_share = 100
+		/// In other cases, the custom recipient_share can be specified
 		#[require_transactional]
-		fn release_payment(from: T::AccountId, to: T::AccountId) -> DispatchResult {
-			use PaymentState::*;
-			// add the payment detail to storage
-			Payment::<T>::try_mutate(
-				from.clone(),
-				to.clone(),
-				|maybe_payment| -> DispatchResult {
-					let payment = maybe_payment.as_mut().ok_or(Error::<T>::InvalidPayment)?;
-					// ensure the payment is in created state
-					ensure!(payment.state == Created, Error::<T>::PaymentAlreadyReleased);
-
-					match &payment.fee_detail {
-						Some((fee_recipient_account, fee_amount)) => {
-							// unreserve the incentive amount + fees back to the creator
-							T::Asset::unreserve(
-								payment.asset,
-								&from,
-								payment.incentive_amount + *fee_amount,
-							);
-							// unreserve the amount to the recipent
-							T::Asset::unreserve(payment.asset, &to, payment.amount);
-							// transfer fee amount to marketplace
-							T::Asset::transfer(
-								payment.asset,
-								&from,                  // fee is paid by payment creator
-								&fee_recipient_account, // account of fee recipient
-								*fee_amount,            // amount of fee
-							)?;
-						},
-						None => {
-							// unreserve the incentive amount back to the creator
-							T::Asset::unreserve(payment.asset, &from, payment.incentive_amount);
-							// unreserve the amount to the recipent
-							T::Asset::unreserve(payment.asset, &to, payment.amount);
-						},
-					}
-
-					// clear payment data from storage
-					*maybe_payment = None;
-					Ok(())
-				},
-			)?;
-			Self::deposit_event(Event::PaymentReleased { from, to });
-			Ok(())
-		}
-
-		/// This function will allows user to cancel a payment. When cancelling a payment, steps are
-		/// - Unreserve the incentive amount
-		/// - Unreserve the payment amount
-		/// - Transfer amount from recipent to sender
-		#[require_transactional]
-		fn cancel_payment(from: T::AccountId, to: T::AccountId) -> DispatchResult {
-			// add the payment detail to storage
+		fn settle_payment(
+			from: T::AccountId,
+			to: T::AccountId,
+			recipient_share: Percent,
+		) -> DispatchResult {
 			Payment::<T>::try_mutate(
 				from.clone(),
 				to.clone(),
 				|maybe_payment| -> DispatchResult {
 					let payment = maybe_payment.take().ok_or(Error::<T>::InvalidPayment)?;
 
-					// unreserve the incentive amount from the owner account
+					// unreserve the incentive amount and fees from the owner account
 					match payment.fee_detail {
-						Some((_, fee_amount)) => {
+						Some((fee_recipient, fee_amount)) => {
 							T::Asset::unreserve(
 								payment.asset,
 								&from,
 								payment.incentive_amount + fee_amount,
 							);
+							// if the settlement is a release, transfer fee to marketplace
```

Comment:
I think marketplace should get the payment fee always, or at least when the share is not 0

Reply:
also let's not mention marketplace since this pallet doesn't know what that is